Improvements to ViewSet extra actions

rpkilby opened the pull request:
Resolves #2062, #2934, #3271, #5755
Fixes #4920

This will obviously require more extensive testing and documentation updates, but I'd like to hammer out the implementation details first. But first, here's a preview of the changes to the browsable API:

![change-password](https://user-images.githubusercontent.com/2370209/32976956-1ca9ab7e-cbf1-11e7-981a-a20cb1e83d63.png)

The above changes include:
- Breadcrumb and page title support custom names/titles for actions instead of just "User instance".
- The method docstring is now used as the action's description.
- A dropdown of "Extra Actions" (appropriately filtered to detail/non-detail actions).
- Also, note that custom serializer classes were already supported.

----

To implement this, it was necessary to change how routers and viewsets interact. At a higher level: 

- `View`s more explicitly customize their display name and description.
- `ViewSet`s are provided more context/configuration from the router. For example, it can now determine if the current request is operation on a list or detail action.

In more detail detail:

- Note that these changes build off of #5648 and #5705, which were spun out of this PR.
- `View`s now accept `name` and `description` via the `initkwargs`. 
    - This provides a general hook for customizing their display on the browsable API.
    - This is necessary to allow action methods to specify they're own name/description.
- `get_view_name` and `get_view_description` now expect the View instance instead of a class. 
- `get_view_name` and `get_view_description` respect the new View `name` and `description`.
- `ViewSet`s consider `name` and `suffix` to be mutually exclusive.
- The `action` decorator now allows users to specify the `name`, uses the docstring as the description.
- Adds `ViewSet.get_extra_action_url_map` for rendering extra actions.
- Adds renderer/template code for both the browsable and admin APIs.
- Adds method mapping to ViewSet actions. Multiple ViewSet methods can handle different HTTP methods of an action. ex:
    ```python
    class MyViewSet(ViewSet):

        @action(detail=False)
        def example(self, request, **kwargs):
            ...

        @example.mapping.post
        def create_example(self, request, **kwargs):
            ...
    ```

**TODO:**
- [x] Look into: #4920, ~~#5630~~
- [x] Add tests for new functionality
- [x] Add actions list to browsable API
- [x] Add actions list to admin API
- [x] Fix action title/description display in browsable API
- [ ] Update docs, changelog
- [ ] ~~Rename/deprecate `base_name` to `basename` for consistency.~~

rpkilby commented:
> I like the naming of action rater then detail/list route.

Agreed @auvipy. It's not so much that I'm in favor of `action`, more that I've never particularly liked `list_route`/`detail_route`.

> how many routable actions will be possible on a single viewset?

There isn't any limit, as there wasn't any limit to list/detail routes.

SIde note: more commits incoming

auvipy commented:
It would be great to see some good example as part of this pr. thanks for the great work!

rpkilby commented:
@auvipy, right now, there's a screenshot in the original comment, which should demonstrate the practical changes to the browsable API. As far as code goes, it's pretty much just:

```python
class MyViewSet(...):
    @action(methods=['get'], detail=True, name='Custom name'
            serializer_class=ThingySerializer)
    def do_a_thing(self, request, pk, *args, **kwargs):
        """
        # Docstrings are now used as descriptions!
        """
```

auvipy commented:
How this change would affect this https://github.com/encode/django-rest-framework/issues/5551?

rpkilby commented:
These changes aren't directly related to #5551.

merwok commented on rest_framework/decorators.py:
```diff
@@ -130,29 +131,51 @@ def decorator(func):
+    warnings.warn(
+        "`detail_route` has been deprecated in favor of `action`, which accepts a "
+        "`detail` boolean. Use `@action(['GET'], detail=True)` instead.",
+        PendingDeprecationWarning
```
I think this needs a `stacklevel` parameter so that the source code in the traceback points to the file using detail_route, not this file.

carltongibson commented on rest_framework/decorators.py:
```diff
@@ -130,29 +131,51 @@ def decorator(func):
+    Used to mark a method on a ViewSet that should be routed for detail requests.
+    """
+    warnings.warn(
+        "`detail_route` has been deprecated in favor of `action`, which accepts a "
```
This is quite a big change for users. I think I'd like to see the message here saying, "pending deprecation" and "will be removed in 3.10" so that it's mega-clear what sort of timescale users have. 

* This will require updating these messages for 3.9 (unless we're clever with the wording) but I think we need to be extra helpful here. 
* I think this also merits a beginning on the release notes.

rpkilby replied:
Agreed. My main consideration right now is to determine if this is the path we ultimately want to move toward, and then work backwards from that w/ deprecations and whatnot. 

Right now, the state of the PR is more exploratory. For example, I also want to try and fix #4920.

carltongibson commented:
Hey @rpkilby. Any appetite for pushing this over the line? It looks like it just needs docs and a few test cases

carltongibson commented:
Nice work @rpkilby. Looks good already. I'll sit down with it over the next week.

rpkilby commented:
Thanks. Also, feel free to nitpick over naming and such. I kept waffling between `mapping` and `methods`, but chose the former.

maryokhin commented:
This is impressive work, something I've always wanted in DRF as an old-time user 👍

rpkilby commented:
Okay, added some documentation. This includes:
- Updated the `VIEW_NAME_FUNCTION`/`VIEW_DESCRIPTION_FUNCTION` docs to the new signature, and documented the relevant `ViewSet` optional arguments (`name`, `suffix`, `detail`, `description`).
- Document the new method mapping enhancement for the `action` decorator.

noisy commented:
I just found this pull request, because I tried to make my actions visible in browsable api. Thank you for this work.

Right now I am developing my project installing DRF directly from git. @carltongibson - do you know when we can expect a new version of DRF with this feature included?

rpkilby commented:
Brief aside: I know there is a little more work that needs to be done before a 3.9 release can be shipped, but one minor blocker is #6081, which fixes a flaw in this pull request. I *should* have time to work on it this weekend.
